### ISSUE TITLE
Record Collection incorrect solution passed all tests

### ISSUE THREAD

**Issue by UtkarshShukla7** [FreeCodeCamp/FreeCodeCamp#8979]
Challenge [Record Collection](https://www.freecodecamp.com/challenges/record-collection#?solution=%0A%2F%2F%20Setup%0Avar%20collection%20%3D%20%7B%0A%20%20%20%202548%3A%20%7B%0A%20%20%20%20%20%20album%3A%20%22Slippery%20When%20Wet%22%2C%0A%20%20%20%20%20%20artist%3A%20%22Bon%20Jovi%22%2C%0A%20%20%20%20%20%20tracks%3A%20%5B%20%0A%20%20%20%20%20%20%20%20%22Let%20It%20Rock%22%2C%20%0A%20%20%20%20%20%20%20%20%22You%20Give%20Love%20a%20Bad%20Name%22%20%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%7D%2C%0A%20%20%20%202468%3A%20%7B%0A%20%20%20%20%20%20album%3A%20%221999%22%2C%0A%20%20%20%20%20%20artist%3A%20%22Prince%22%2C%0A%20%20%20%20%20%20tracks%3A%20%5B%20%0A%20%20%20%20%20%20%20%20%221999%22%2C%20%0A%20%20%20%20%20%20%20%20%22Little%20Red%20Corvette%22%20%0A%20%20%20%20%20%20%5D%0A%20%20%20%20%7D%2C%0A%20%20%20%201245%3A%20%7B%0A%20%20%20%20%20%20artist%3A%20%22Robert%20Palmer%22%2C%0A%20%20%20%20%20%20tracks%3A%20%5B%20%5D%0A%20%20%20%20%7D%2C%0A%20%20%20%205439%3A%20%7B%0A%20%20%20%20%20%20album%3A%20%22ABBA%20Gold%22%0A%20%20%20%20%7D%0A%7D%3B%0A%2F%2F%20Keep%20a%20copy%20of%20the%20collection%20for%20tests%0Avar%20collectionCopy%20%3D%20JSON.parse(JSON.stringify(collection))%3B%0A%0A%2F%2F%20Only%20change%20code%20below%20this%20line%0A%0Afunction%20updateRecords(id%2C%20prop%2C%20value)%20%7B%0A%20%20if%20(value%20!%3D%3D%20%22%22%20%26%26%20prop%20!%3D%3D%20%22tracks%22)%20%7B%0A%20%20%20%20collection%5Bid%5D%5Bprop%5D%20%3D%20value%3B%0A%20%20%7D%20else%20if%20(prop%20%3D%3D%3D%20%22tracks%22%20%26%26%20value%20!%3D%3D%20%22%22)%20%7B%0A%20%20%20%20%2F*if(collection%5Bid%5D.hasOwnProperty(prop))%0A%20%20%20%20%20%20collection%5Bid%5D%5Bprop%5D.push(value)%3B%0A%20%20%20%20else%0A%20%20%20%20%20%20collection%5Bid%5D%5Bprop%5D%3D%5Bvalue%5D%3B%0A%20%20%20%20%2F%2Fcollection%5Bid%5D%5Bprop%5D.push(value)%3B*%2F%0A%20%20%20%20collection%5Bid%5D%5Bprop%5D%3D%5Bvalue%5D%3B%0A%20%20%7D%20else%20if%20(value%20%3D%3D%3D%20%22%22)%20%7B%0A%20%20%20delete%20collection%5Bid%5D%5Bprop%5D%3B%0A%20%20%7D%0A%0A%20%20return%20collection%3B%0A%7D%0A%0A%0A%0A%2F%2F%20Alter%20values%20below%20to%20test%20your%20code%0AupdateRecords(2548%2C%20%22tracks%22%2C%20%22ABBA%22)%3B%0A%0A) has an issue.
User Agent is: <code>Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36</code>.
Please describe how to reproduce this issue, and include links to screenshots if possible.

My code:
```javascript

// Setup
var collection = {
    2548: {
      album: "Slippery When Wet",
      artist: "Bon Jovi",
      tracks: [ 
        "Let It Rock", 
        "You Give Love a Bad Name" 
      ]
    },
    2468: {
      album: "1999",
      artist: "Prince",
      tracks: [ 
        "1999", 
        "Little Red Corvette" 
      ]
    },
    1245: {
      artist: "Robert Palmer",
      tracks: [ ]
    },
    5439: {
      album: "ABBA Gold"
    }
};
// Keep a copy of the collection for tests
var collectionCopy = JSON.parse(JSON.stringify(collection));

// Only change code below this line

function updateRecords(id, prop, value) {
  if (value !== "" && prop !== "tracks") {
    collection[id][prop] = value;
  } else if (prop === "tracks" && value !== "") {
    /*if(collection[id].hasOwnProperty(prop))
      collection[id][prop].push(value);
    else
      collection[id][prop]=[value];
    //collection[id][prop].push(value);*/
    collection[id][prop]=[value];
  } else if (value === "") {
   delete collection[id][prop];
  }

  return collection;
}



// Alter values below to test your code
updateRecords(2548, "tracks", "ABBA");


```

This shouldn't work because its redefining tracks instead of pushing value to it (in the case when value isn't null).But it passed all tests .So ,you should not only just check that the last element of tracks should be val but also that the previous array elements are also present in it or maybe check that the size of tracks is increased and isn't reduced/unaltered (if tracks property is present).

**Comment by eljaygalvez** [comment]
Hi,
I also got stuck in this challenge. 

Here's my code: 
`
// Setup
var collection = {
    2548: {
      album: "Slippery When Wet",
      artist: "Bon Jovi",
      tracks: [ 
        "Let It Rock", 
        "You Give Love a Bad Name" 
      ]
    },
    2468: {
      album: "1999",
      artist: "Prince",
      tracks: [ 
        "1999", 
        "Little Red Corvette" 
      ]
    },
    1245: {
      artist: "Robert Palmer",
      tracks: [ ]
    },
    5439: {
      album: "ABBA Gold"
    }
};
// Keep a copy of the collection for tests
var collectionCopy = JSON.parse(JSON.stringify(collection));

// Only change code below this line
function updateRecords(id, prop, value) {
  
  if(value !== "") {
      if(prop === "tracks") { 
        collection[id][prop].push(value);
    } else
        collection[id][prop] = value;
  } else {
     delete collection[id][prop];
  }

  return collection;
}

// Alter values below to test your code
updateRecords(5439, "tracks", "Take a Chance on Me");

`
I still don't understand why it is showing me this, 
> TypeError: Cannot read property 'push' of undefined.
Shouldn't we use .push to add the value onto the end of the array. It got me confused.

But for now, I'll copy your code so I could move on. haha.

**Comment by UtkarshShukla7** [comment]
@eljaygalvez you got that error because in the collection object the object with id 5439 doesn't have tracks property predefined .So when you try to push into the tracks property (which is undefined currently for this object) ,the push operation cannot be performed as it is only defined for arrays.If you will see the commented part in my solution ,you will be able to find the correct solution

**Comment by xboudsady** [comment]
@UtkarshShukla7 Thanks man for the explanation.  This was so frustrating, I think every other solution using google search is not correct for this challenge update. As you said, it's "re-defining" track instead of pushing it?

**Comment by thoragio** [comment]
@UtkarshShukla7 Well spotted. This is truly an issue in the code checking routine.

**Comment by erictleung** [comment]
@UtkarshShukla7 

> So ,you should not only just check that the last element of tracks should be val but also that the previous array elements are also present in it or maybe check that the size of tracks is increased and isn't reduced/unaltered (if tracks property is present).

Great point. There should be some additional tests that check for the pre-existing elements in the arrays after new ones have been pushed to it.

**Comment by UtkarshShukla7** [comment]
@erictleung Thanks for the review ! Also ,one test can be added to check if `tracks` property is an array before pushing into it but I guess we can let it go as we know how our object looks like and the description states that you have to push value into it (which tells that tracks would always be an array, if present).

**Comment by erictleung** [comment]
@UtkarshShukla7 the pushing to an non-existent array issue should be resolved soon with #9034 once it's pushed to the live site.

**Comment by UtkarshShukla7** [comment]
@erictleung That seems to be covered then . :+1:
